topdown/bytes.go

```diff
@@ -0,0 +1,95 @@
+package topdown
+
+import (
+	"fmt"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+	"math/big"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type multiplier int
+
+const (
+	base = 1024
+
+	none multiplier = 1
+	kb              = base
+	mb              = kb * base
+	gb              = mb * base
+	tb              = gb * base
+)
+
+func numBytesErr(msg string) error {
+	return fmt.Errorf("numbytes error: %s", msg)
+}
+
+var (
+	errNoAmount = numBytesErr("no byte amount provided")
+	errIntConv  = numBytesErr("could not parse amount to integer")
+)
+
+func errUnitNotRecognized(unit string) error {
+	return numBytesErr(fmt.Sprintf("unit %s not recognized", unit))
+}
+
+func builtinNumBytes(a ast.Value) (ast.Value, error) {
+	var m multiplier
+
+	numPattern := `([0-9]+)`
+	strPattern := `([a-z]+)`
+
+	raw, err := builtins.StringOperand(a, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	s := strings.ToLower(raw.String())
+
+	numRe, err := regexp.Compile(numPattern)
+	if err != nil {
+		return nil, err
+	}
+
+	strRe, err := regexp.Compile(strPattern)
+	if err != nil {
+		return nil, err
+	}
+
+	numStr := numRe.FindString(s)
+	unit := strRe.FindString(s)
+
+	if numStr == "" {
+		return nil, errNoAmount
+	}
+
+	num, err := strconv.Atoi(numStr)
+	if err != nil {
+		return nil, errIntConv
+	}
+
+	switch unit {
+	case "":
+		m = none
+	case "k", "kb", "ki":
+		m = kb
+	case "m", "mb", "mi":
+		m = mb
+	case "g", "gb", "gi":
+		m = gb
+	case "t", "tb", "ti":
+		m = tb
+	default:
+		return nil, errUnitNotRecognized(unit)
```

Comment:
@tsandall what are you're thoughts on an error message here vs making it undefined? I'm not 100% sure on the mechanics to do that but I know its something we've talked about for builtins.. seems like this might be a good place to do it.

My thinking is that some policy might want to not just explode if the string is malformed and instead have a rule to handle that case with some default or better error message (or both).

Reply:
Update: I chatted with @tsandall and we don't want to complicate things by doing the undefined thing now. Lets stick with the error for now.

Reply:
@patrick-east Okay, cool. The actual error message is a little off the cuff so let me know if you prefer a language change there.